azext_edge/edge/providers/orchestration/upgrade.py

```diff
@@ -0,0 +1,295 @@
+from rich.live import Live
+from rich.padding import Padding
+from rich.progress import Progress, SpinnerColumn, TimeElapsedColumn
+from rich.table import Table
+
+from ...util.az_client import get_resource_client, wait_for_terminal_state
+from ...util.common import should_continue_prompt
+from .resource_map import IoTOperationsResourceMap
+from .resources import Instances
+
+logger = get_logger(__name__)
+INSTANCE_7_API = "2024-08-15-preview"
+
+
+def upgrade_ops_resources(
+    cmd,
+    resource_group_name: str,
+    instance_name: Optional[str] = None,
+    cluster_name: Optional[str] = None,
+    sr_resource_id: Optional[str] = None,
+    confirm_yes: Optional[bool] = None,
+    no_progress: Optional[bool] = None,
+):
+    manager = UpgradeManager(
+        cmd=cmd,
+        instance_name=instance_name,
+        cluster_name=cluster_name,
+        sr_resource_id=sr_resource_id,
+        resource_group_name=resource_group_name,
+        no_progress=no_progress,
+    )
+    return manager.do_work(confirm_yes=confirm_yes)
+
+
+# keeping this separate for easier removal once no longer needed
+class UpgradeManager:
+    def __init__(
+        self,
+        cmd,
+        resource_group_name: str,
+        instance_name: Optional[str] = None,
+        cluster_name: Optional[str] = None,
+        sr_resource_id: Optional[str] = None,
+        no_progress: Optional[bool] = None,
+    ):
+        from azure.cli.core.commands.client_factory import get_subscription_id
+
+        self.cmd = cmd
+        self.instance_name = instance_name
+        self.cluster_name = cluster_name
+        self.sr_resource_id = sr_resource_id
+        self.resource_group_name = resource_group_name
+        self.instances = Instances(self.cmd)
+        self.subscription_id = get_subscription_id(cli_ctx=cmd.cli_ctx)
+        self.resource_client = get_resource_client(self.subscription_id)
+
+        self._render_progress = not no_progress
+        self._live = Live(None, transient=False, refresh_per_second=8, auto_refresh=self._render_progress)
+        self._progress_bar = Progress(
+            SpinnerColumn(),
+            *Progress.get_default_columns(),
+            "Elapsed:",
+            TimeElapsedColumn(),
+            transient=False,
+        )
+        self._progress_shown = False
+
+    def do_work(self, confirm_yes: Optional[bool] = None):
+        # get the resource map from the instance (checks if update is needed for instance)
+        self.resource_map = self._get_resource_map()
+        # Ensure cluster exists with existing resource_map pattern.
+        self.resource_map.connected_cluster.resource
+        self.cluster_name = self.resource_map.connected_cluster.cluster_name
+
+        # get the extensions to update, populate the expected patches
+        extension_text = self._check_extensions()
+
+        if not self.extensions_to_update and not self.require_instance_upgrade:
+            logger.warning("Nothing to upgrade :)")
```

Comment:
why warning and not info?

Reply:
you know what? screw it I like print better